core/src/main/java/org/elasticsearch/discovery/zen/ping/ZenPingService.java

```diff
@@ -78,60 +75,31 @@ protected void doClose() {
         }
     }
 
-    public PingResponse[] pingAndWait(TimeValue timeout) {
-        final AtomicReference<PingResponse[]> response = new AtomicReference<>();
-        final CountDownLatch latch = new CountDownLatch(1);
-        ping(new PingListener() {
-            @Override
-            public void onPing(PingResponse[] pings) {
-                response.set(pings);
-                latch.countDown();
+    public ZenPing.PingCollection pingAndWait(TimeValue timeout) {
+        final ZenPing.PingCollection response = new ZenPing.PingCollection();
+        final CountDownLatch latch = new CountDownLatch(zenPings.size());
+        for (ZenPing zenPing : zenPings) {
+            final AtomicBoolean counted = new AtomicBoolean();
+            try {
+                zenPing.ping(pings -> {
+                    response.addPings(pings);
```

Comment:
tja - doesn't really matter. I figured every extra bit of information, if we manage to get it in, counts